##### Задания не несут смысловой нагрузки для реальной жизни, они сделаны лишь для того, чтоб научить использовать Pandas

This database contains 76 attributes, but all published experiments refer to using a subset of 14 of them. In particular, the Cleveland database is the only one that has been used by ML researchers to this date. The "goal" field refers to the presence of heart disease in the patient. It is integer valued from 0 (no presence) to 1.

https://www.kaggle.com/ronitf/heart-disease-uci

___

In [1]:
import pandas as pd
import numpy as np

In [2]:
data = pd.read_csv('heart_pandas.csv')
data

age     sex   chest_pain_type  resting_blood_pressure  cholesterol  \
0     63    male  non-anginal pain                     145          233   
1     37    male   atypical angina                     130          250   
2     41  female    typical angina                     130          204   
3     56    male    typical angina                     120          236   
4     57  female    typical angina                     120          354   
..   ...     ...               ...                     ...          ...   
298   57  female    typical angina                     140          241   
299   45    male  non-anginal pain                     110          264   
300   68    male    typical angina                     144          193   
301   57    male    typical angina                     130          131   
302   57  female    typical angina                     130          236   

       fasting_blood_sugar               rest_ecg  max_heart_rate_achieved  \
0    greater than 120mg/ml                 normal                      150   
1      lower than 120mg/ml  ST-T wave abnormality                      187   
2      lower than 120mg/ml                 normal                      172   
3      lower than 120mg/ml  ST-T wave abnormality                      178   
4      lower than 120mg/ml  ST-T wave abnormality                      163   
..                     ...                    ...                      ...   
298    lower than 120mg/ml  ST-T wave abnormality                      123   
299    lower than 120mg/ml  ST-T wave abnormality                      132   
300  greater than 120mg/ml  ST-T wave abnormality                      141   
301    lower than 120mg/ml  ST-T wave abnormality                      115   
302    lower than 120mg/ml                 normal                      174   

    exercise_induced_angina  st_depression   st_slope  num_major_vessels  \
0                        no            2.3  upsloping                  0   
1                        no            3.5  upsloping                  0   
2                        no            1.4       flat                  0   
3                        no            0.8       flat                  0   
4                       yes            0.6       flat                  0   
..                      ...            ...        ...                ...   
298                     yes            0.2  upsloping                  0   
299                      no            1.2  upsloping                  0   
300                      no            3.4  upsloping                  2   
301                     yes            1.2  upsloping                  1   
302                      no            0.0  upsloping                  1   

           thalassemia  target  
0               normal       1  
1         fixed defect       1  
2         fixed defect       1  
3         fixed defect       1  
4         fixed defect       1  
..                 ...     ...  
298  reversable defect       0  
299  reversable defect       0  
300  reversable defect       0  
301  reversable defect       0  
302       fixed defect       0  

[303 rows x 14 columns]

#### Описание признаков

**age**: The person's age in years

**sex**: The person's sex (1 = male, 0 = female)

**cp**: The chest pain experienced (Value 1: typical angina, Value 2: atypical angina, Value 3: non-anginal pain, Value 4: asymptomatic)

**trestbps**: The person's resting blood pressure (mm Hg on admission to the hospital)

**chol**: The person's cholesterol measurement in mg/dl

**fbs**: The person's fasting blood sugar (> 120 mg/dl, 1 = true; 0 = false)

**restecg**: Resting electrocardiographic measurement (0 = normal, 1 = having ST-T wave abnormality, 2 = showing probable or definite left ventricular hypertrophy by Estes' criteria)

**thalach**: The person's maximum heart rate achieved

**exang**: Exercise induced angina (1 = yes; 0 = no)

**oldpeak**: ST depression induced by exercise relative to rest ('ST' relates to positions on the ECG plot. See more here)

**slope**: the slope of the peak exercise ST segment (Value 1: upsloping, Value 2: flat, Value 3: downsloping)

**ca**: The number of major vessels (0-3)

**thal**: A blood disorder called thalassemia (3 = normal; 6 = fixed defect; 7 = reversable defect)

**target**: Heart disease (0 = no, 1 = yes)

> 1. age 
> 2. sex 
> 3. chest pain type (4 values) 
> 4. resting blood pressure 
> 5. serum cholestoral in mg/dl 
> 6. fasting blood sugar > 120 mg/dl
> 7. resting electrocardiographic results (values 0,1,2)
> 8. maximum heart rate achieved 
> 9. exercise induced angina 
> 10. oldpeak = ST depression induced by exercise relative to rest 
> 11. the slope of the peak exercise ST segment 
> 12. number of major vessels (0-3) colored by flourosopy 
> 13. thal: 3 = normal; 6 = fixed defect; 7 = reversable defect

___

#### Сколько мужчин в датасете? Сколько женщен? (sex)

In [7]:
## Ваш код
data.sex.value_counts()

sex
male      207
female     96
Name: count, dtype: int64

#### Какой процент мужчин в датасете? (решите в одну строчку, не используя предыдущий результат. Не забудте знак процента) 

In [56]:
## Ваш код
male_percent = data.sex[data.sex == 'male'].count() / data.sex.count() * 100
print(f'{round(male_percent,1)} %')

68.3 %


#### Сколько мужчин имеют заболевание сердца? Сколько женщин имеют заболевание сердца?

In [27]:
## Ваш код
heart_d = data.groupby(data.target[data.target == 1])
heart_d.sex.value_counts()


target  sex   
1.0     male      93
        female    72
Name: count, dtype: int64

**Какую долю, от общего числа пациентов, занимают мужчины не имеющие болезнь сердца?**

In [77]:
## Ваш код
healty_male = data.sex[data.sex == 'male'][data.target == 0].count() / data.sex.count()
print(f'Доля мужчин не страдающих сердечными заболеваниями - {round(healty_male, 2)}')



Доля мужчин не страдающих сердечными заболеваниями - 0.38


**Сколько лет самому молодому пациенту, страдающему болезнью сердца?**

In [63]:
## Ваш код
data.age[data.target == 1].min()

29

**Сколько лет самому возастному пациенту, у которого нет проблем с сердцем?**

In [66]:
## Ваш код
data.age[data.target == 0].max()

77

**Сколько лет самой молодой женщине, которая страдает болезнью сердца?**

In [71]:
## Ваш код
data.age[data.sex == 'female'][data.target == 1].min()

34

**Какой средний возраст женщин?**

In [73]:
## Ваш код
round(data.age[data.sex == 'female'].mean(), 1)

55.7

**Каковы средние значения и среднеквадратичные отклонения возраста тех, кто страдают болезнью сердца?**

In [82]:
## Ваш код
heart_d_m = data.groupby(data.target[data.target == 1])
print(round(heart_d_m.age.std(), 1), round(heart_d_m.age.mean(), 1))

target
1.0    9.6
Name: age, dtype: float64 target
1.0    52.5
Name: age, dtype: float64


#### Правда ли, что люди не болеющие болезнью сердца имеют уровня холестерина меньше среднего? (chol) 

In [87]:
## Ваш код
gen_m = data.cholesterol.mean()
heart_d_m = data.cholesterol[data.target == 1].mean()
result = gen_m > heart_d_m
result

True

**Выведите статистику rest_ecg для все числовых признаков, его максимально и среднее значение (используйте groupby(), решите в одну строчку)**

In [ ]:
## Ваш код
data.groupby(data.rest_ecg).describe().loc[['mean', 'max']]


age                                           \
                              count       mean        std   min    25%   50%   
rest_ecg                                                                       
ST-T wave abnormality         152.0  52.914474   9.229196  34.0  44.75  53.5   
left ventricular hypertrophy    4.0  61.000000  10.099505  55.0  55.00  56.5   
normal                        147.0  55.687075   8.675902  29.0  50.50  57.0   

                                         resting_blood_pressure              \
                               75%   max                  count        mean   
rest_ecg                                                                      
ST-T wave abnormality         60.0  71.0                  152.0  129.065789   
left ventricular hypertrophy  62.5  76.0                    4.0  140.500000   
normal                        62.0  77.0                  147.0  134.027211   

                              ... num_major_vessels      target            \
                              ...               75%  max  count      mean   
rest_ecg                      ...                                           
ST-T wave abnormality         ...               1.0  4.0  152.0  0.631579   
left ventricular hypertrophy  ...               1.5  3.0    4.0  0.250000   
normal                        ...               1.0  4.0  147.0  0.462585   

                                                                  
                                   std  min  25%  50%   75%  max  
rest_ecg                                                          
ST-T wave abnormality         0.483971  0.0  0.0  1.0  1.00  1.0  
left ventricular hypertrophy  0.500000  0.0  0.0  0.0  0.25  1.0  
normal                        0.500303  0.0  0.0  0.0  1.00  1.0  

[3 rows x 56 columns]

**Посчитайте у кого уровень депрессии при физический нагрузке выше (в среднем), среди мужчин страдающих болезнью сердца или среди женщин не страдающих болезнью сердца (st_depression)**

In [89]:
## Ваш код
male_unhealthy = data.st_depression[data.sex == 'male'][data.target == 1].mean()
female_healthy = data.st_depression[data.sex == 'female'][data.target == 0].mean()
if male_unhealthy > female_healthy:
    print(f'Уровень депресии у мужчин выше - {round(male_unhealthy, 1)}')
else:
    print(f'Уровень депресии у женшин выше - {round(female_healthy, 1)}')


Уровень депресии у женшин выше - 1.8


**Посчтитайте максимальный и минимальный уровень холестерина для каждого типа chest_pain_type, rest_ecg, thalassemia. Пишите код оптимально, можно использовать циклы**

In [1]:
## Ваш код
chest_pain = data.groupby([data.chest_pain_type, data.cholesterol]).describe().loc[['min', 'max']]
rest_ecg = data.groupby([data.rest_ecg, data.cholesterol]).describe().loc[['min', 'max']]
thalassemia = data.groupby([data.thalassemia, data.cholesterol]).describe().loc[['min', 'max']]



NameError: name 'data' is not defined

**Сколько значений может принимать каждый из категориальных признаков?**

In [16]:
## Ваш код

**У какого категориального признака наблюдается самый сильный дизбаланс классов?**

In [17]:
## Ваш код

---